### PR TITLE
Added mypy type hinting to hooks, worker, encoding, payload, pin, sampler

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
Following #2180, the following files were type hinted for mypy:
- ddtrace/_hooks.py
- ddtrace/_worker.py
- ddtrace/encoding.py
- ddtrace/pin.py
- ddtrace/payload.py
- ddtrace/sampler.py

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/_hooks.py [hunk 1]
```diff
@@ -29,7 +32,11 @@ def __deepcopy__(self, memodict=None):
         hooks._hooks = deepcopy(self._hooks, memodict)
         return hooks
 
-    def register(self, hook, func=None):
+    def register(
+        self,
+        hook,  # type: Any
+        func=None,  # type: Optional[Callable]
+    ):
         """
```
Comment (P403n1x87) [inline ddtrace/_hooks.py:40]: return type?

File: ddtrace/sampler.py [hunk 2]
```diff
@@ -78,24 +91,39 @@ class RateByServiceSampler(BaseSampler, BasePrioritySampler):
...
+    ):
+        # type: (...) -> str
         """Compute a key with the same format used by the Datadog agent API."""
         service = service or ""
         env = env or ""
         return "service:" + service + ",env:" + env
 
-    def __init__(self, sample_rate=1):
+    def __init__(
+        self, sample_rate=1  # type: float
```
Comment (P403n1x87) [inline ddtrace/sampler.py:105]:
```suggestion
        self, sample_rate=1.0  # type: float
```

File: ddtrace/internal/_encoding.pyi [hunk 3]
```diff
@@ -1,6 +1,9 @@
 from typing import Any
+from typing import List
 from typing import Union
 
 class MsgpackEncoder(object):
     content_type: str
     def _decode(self, data: Union[str, bytes]) -> Any: ...
+    def encode_trace(self, trace: List[Any]) -> bytes: ...
+    def join_encoded(self, objs: List[bytes]) -> Union[str, bytes]: ...
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/internal/_encoding.pyi:9]: Ahh I assumed it should have the same return type as the input type for `_decode`, but makes sense I'll change that!

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — DISMISSED** [review]

File: ddtrace/sampler.py [hunk 4]
```diff
@@ -52,7 +62,8 @@ class RateSampler(BaseSampler):
     It samples randomly, its main purpose is to reduce the instrumentation footprint.
     """
 
-    def __init__(self, sample_rate=1):
+    def __init__(self, sample_rate=1.0):
```
Comment (Kyle-Verhoog) [inline ddtrace/sampler.py:65]: Can we do this change separately? There are further usages below that use `int`s

File: ddtrace/sampler.py [hunk 5]
```diff
@@ -52,7 +62,8 @@ class RateSampler(BaseSampler):
...
+        # type: (float) -> None
         if sample_rate < 0:
             raise ValueError("sample_rate of {} is negative".format(sample_rate))
         elif sample_rate > 1:
```
Comment (Kyle-Verhoog) [inline ddtrace/sampler.py:69]: This needs to go away 🤦‍♂️

Comment (Yun-Kim) [reply ddtrace/sampler.py:69]: Do you mean the type hint or changing the default value to be a float?

File: ddtrace/sampler.py (hunk 4 above)
Comment (Yun-Kim) [reply ddtrace/sampler.py:65]: Sounds fair to me, I'll address this in a separate PR then 👍

File: ddtrace/sampler.py (hunk 5 above)
Comment (Kyle-Verhoog) [reply ddtrace/sampler.py:69]: lol nothing for this PR, just me cringing at this code that needs to change

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by jd — APPROVED** [review]

**Pull request merged**